### PR TITLE
Task 4 - Integration with Database

### PR DESCRIPTION
**Pull request by abachyla**
1. [Task](https://github.com/EPAM-JS-Competency-center/cloud-development-course-initial/blob/main/4_integration_with_database/task.md)
2. Links
`getProductsList`: GET - https://kun8h51167.execute-api.eu-west-1.amazonaws.com/products
`getProductsById`:  GET - https://kun8h51167.execute-api.eu-west-1.amazonaws.com/products/{productId}
`createProduct`:  POST - https://kun8h51167.execute-api.eu-west-1.amazonaws.com/products
CloudFront: https://d14axewerpas2.cloudfront.net/

4. FE PR: https://github.com/EllenDanilowa/shop-angular-cloudfront/pull/4
5. Done 14.07.2022 / Deadline 24.07.2022
6. Score 9/9
- [x] **1** - Task 4.1 is implemented
- [x] **3** - Task 4.2 is implemented lambda links are provided and returns data
- [x] **4** - Task 4.3 is implemented lambda links are provided and products is stored in DB (call Task 4.2 to see the product)
- [x] **5** - Your own Frontend application is integrated with Product Service (`/products` API) and products from Product Service are represented on Frontend. Link to a working Frontend application is provided for cross-check reviewer.

Additional:
- [x] **+1** **(All languages)** - POST `/products` lambda functions returns error 400 status code if product data is invalid
- [x] **+1** **(All languages)** - All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
- [x] **+1** **(All languages)** - All lambdas do `console.log` for each incoming requests and their arguments
- [x] **+1** **(All languages)** - Transaction based creation of product (in case stock creation is failed then related to this stock product is not created and not ready to be used by the end user and vice versa) (https://devcenter.kinvey.com/nodejs/tutorials/bl-transactional-support)

### CODE REVIEW

**Comment by yulei-chen** [comment]
Hi @EllenDanilowa , I can't visit CloudFront: https://d14axewerpas2.cloudfront.net/. It shows 500 error. Could you please check it?

**Comment by abachyla** [comment]
> Hi @EllenDanilowa , I can't visit CloudFront: https://d14axewerpas2.cloudfront.net/. It shows 500 error. Could you please check it?

Hi @yuleicul, sorry for this. I fixed, could you take a look again please?

File: product-service/serverless.yml [hunk 1]
```diff
@@ -12,14 +12,20 @@ provider:
   region: eu-west-1
   httpApi:
     cors: true
+  environment:
+    PG_HOST:
+    PG_PORT:
+    PG_DATABASE:
+    PG_USERNAME:
+    PG_PASSWORD:
```
Comment (SlavaJSFE) [inline product-service/serverless.yml]: How does it works if you didn't set values?

File: product-service/helpers/validateProduct.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    required: true,
+  },
+  count: {
+    type: Number,
+    required: true,
+  },
+});
+
+export const validateProduct = (product) => {
+  const errors = productSchema.validate(product);
```
Comment (SlavaJSFE) [inline product-service/helpers/validateProduct.js:28]: You don't validate on negative integers. I have created products with invalid price and count:
image

**Review by SlavaJSFE — CHANGES_REQUESTED** [review]

File: product-service/handlers/createProduct.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+      throw new Error(ERROR_TYPES.PRODUCT_VALIDATION);
+    }
+
+    const product = await insertProduct(data);
+
+    console.log('Result');
+    console.log(product);
+
+    response = getResponse(RESPONSE_STATUSES.OK, product);
```
Comment (SlavaJSFE) [inline product-service/handlers/createProduct.js]: It's better to return 201 status Code on create action.

**Comment by yulei-chen** [comment]
> > Hi @EllenDanilowa , I can't visit CloudFront: https://d14axewerpas2.cloudfront.net/. It shows 500 error. Could you please check it?
> 
> Hi @yuleicul, sorry for this. I fixed, could you take a look again please?

Checked. No problem.

File: product-service/serverless.yml (hunk 1 above)
Comment (abachyla) [reply product-service/serverless.yml]: Thanks, added dotenv plugin

File: product-service/helpers/validateProduct.js (hunk 2 above)
Comment (abachyla) [reply product-service/helpers/validateProduct.js:28]: Done.

**Review by SlavaJSFE — APPROVED** [review]

**Pull request merged**